serenata_toolbox/presences.py

```diff
@@ -1,143 +1,165 @@
-    print("Fetching data for", filtered_deputies.shape[0], "deputies")
-    for i, deputy in filtered_deputies.iterrows():
-        print(i, deputy['name'])
-        file = urllib.request.urlopen(url.format(deputy['registration']))
-        t = ET.ElementTree(file=file)
-        congressperson_name = extract_text(t.getroot(), 'nomeParlamentar')
-        for day in t.getroot().findall('.//dia'):
-            date = extract_datetime(day, 'data')
+import pandas as pd
+
+from serenata_toolbox import datasets
+from serenata_toolbox.cleanup import xml_extract_text, \
+                                     xml_extract_datetime, \
+                                     xml_extract_date, \
+                                     translate_column
+
+class Presences:
+
+    URL = (
+        'http://www.camara.leg.br/SitCamaraWS/sessoesreunioes.asmx/ListarPresencasParlamentar'
+        '?dataIni={}'
+        '&dataFim={}'
+        '&numMatriculaParlamentar={}'
+    )
+
+    def fetch(self, deputies, start_date, end_date):
+        """
+        :param deputies: (pandas.DataFrame) a dataframe with deputies data
+        :param date_start: (str) date in the format dd/mm/yyyy
+        :param date_end: (str) date in the format dd/mm/yyyy
+        """
+        print("Fetching data for {} deputies from {} -> {}".format(len(deputies), start_date, end_date))
+
+        records = self.__all_presences(deputies, start_date, end_date)
+
+        df = pd.DataFrame(records, columns=(
+            'term',
+            'congressperson_document',
+            # TODO: Rename to something else to avoid confusion (maybe just congressperson?)
+            'congressperson_name',
+            'party',
+            'state',
+            'date',
+            'present_on_day',
+            'justification',
+            'session',
+            'presence'
+        ))
+        return self.__translate(df)
+
+    def __all_presences(self, deputies, start_date, end_date):
+        error_count = 0
+        for i, deputy in deputies.iterrows():
+            print(i, deputy.congressperson_name, deputy.congressperson_document)
+            url = self.URL.format(start_date, end_date, deputy.congressperson_document)
+            xml = self.__try_fetch_xml(10, url)
+
+            if xml == None:
+                error_count += 1
+            else:
+                root = ET.ElementTree(file=xml).getroot()
+                for presence in self.__parse_deputy_presences(root):
+                    yield presence
+
+            time.sleep(2)
+
+        print("\nErrored fetching", error_count, "deputy presences")
+
+    def __try_fetch_xml(self, attempts, url):
+        while attempts > 0:
+            try:
+                return urllib.request.urlopen(url, data=None, timeout=10)
+            except urllib.error.HTTPError as err:
+                print("HTTP Error", err.code, "when loading URL", url)
+                # 500 seems to be the error code for "no data found for the
```

Comment:
Ow boy… I don'r remember where, but another service from the government used to respond `HTTP 200: Arquivo não encontrado` hahaha…